### PR TITLE
Monitoring of Hcal adc sticking in neighbouring TS

### PR DESCRIPTION
**Pull request by zhokin2**
description:
============
The monitoring of possible adc sticking in few neughbouring time-slices (TS) of HBHE subdetectors are implemented as a dedicated button in CMT site. The CMT code is HCAL certification monitoring tool.
In framework of general Remote Monitoring Tool (RMT), the CMT was specially designed to facilitate certification decision on lumi-section(!!!) level. CMT is not integrated in DQM and is used in parallel with DQM, in order to have more HCAL information and for the cross-check of some DQM results.

last presentations related CMT:
https://indico.cern.ch/event/1336438/contributions/5625894/attachments/2733639/4762912/zzzhcal762.pdf
https://indico.cern.ch/event/1336703/contributions/5660532/attachments/2757367/4801293/HCAL_DQM_PPD_Workshop_22112023.pdf
https://indico.cern.ch/event/1349428/contributions/5680674/attachments/2765620/4817262/HCAL_DQM_CMS_Week_05122023.pdf

link to CMT twiki:
https://twiki.cern.ch/twiki/bin/viewauth/CMS/HcalRemoteCertificationMonitoring

CMS sites related to different tasks of general RMT/CMT code:
https://cms-conddb.cern.ch/eosweb/hcal/HcalRemoteMonitoring/RMT
https://cms-conddb.cern.ch/eosweb/hcal/HcalRemoteMonitoring/CMT
https://cms-conddb.cern.ch/eosweb/hcal/HcalRemoteMonitoring/GlobalRMT
https://cms-conddb.cern.ch/eosweb/hcal/HcalRemoteMonitoring/GlobalPSM
https://cms-conddb.cern.ch/eosweb/hcal/HcalRemoteMonitoring/ALCARECOPSM
https://cms-conddb.cern.ch/eosweb/hcal/HcalRemoteMonitoring/IMPSM

the commands were issued:
========================
scram build code-format
scram b checker
scram build code-checks

scram b

cd ../macros/cmt/
./compile.csh RemoteMonitoringGLOBAL.cc &
./compile.csh RemoteMonitoringMAP_Global.cc &
cd ../macros/rmt/
./compile.csh RemoteMonitoringMAP.cc

results:
=======
.....
scan-build: No bugs found
.....
Deleting: No Diagnostics found
.....
clang-format -i src/DPGAnalysis/HcalTools/macros/cmt/RemoteMonitoringGLOBAL.cc
clang-format -i src/DPGAnalysis/HcalTools/macros/cmt/RemoteMonitoringMAP_Global.cc
clang-format -i src/DPGAnalysis/HcalTools/plugins/CMTRawAnalyzer.cc
.....
>> Local Products Rules ..... started
>> Local Products Rules ..... done
.....

### CODE REVIEW

**Comment by zhokin2** [comment]
There is a clarification on the sticking effect implemented in CMT. Monitoring is presented in CMT site as example:

https://cms-conddb.cern.ch/eosweb/hcal/HcalRemoteMonitoring/CMT/GLOBAL_378750/StickingADCcounts_GL.html

The first version of the sticking criterion is when all 8 time slices(TS) stick at the same adc value. Denote the described strict sticking criterion for subdetectors HBHE as HB8 and HE8. If we assume that sticking occurs for a short time and affects only some neighbouring TS, one can consider the case of adc sticking in 6 any neighbouring TS  (HB6 and HE6). Then, consider any 4,3 and even 2 neighbouring sticking (HBHE4, HBHE3, HBHE2). In total there are 5 cases of sticking selection. We construct for all cases the dependence of the number of sticking adc value.  Last plot - the number of adc value w/o any condition applied.

For moment, we select the cases of sticking, along with CapId non rotation,  using the cut on adc amplitude > 135 adc counts in every TS (case HBHE8)

File: DPGAnalysis/HcalTools/macros/cmt/RemoteMonitoringGLOBAL.cc [hunk 1]
```diff
@@ -692,8 +694,8 @@ int main(int argc, char *argv[]) {
         MapNumBadChanDepth[test][sub][k]->Divide(
             MapNumBadChanDepth[test][sub][k], MapNumChanDepth[test][sub][k], 1, 1, "B");
 
-        for (int x = 1; x <= MapNumBadChanFull[test]->GetXaxis()->GetNbins(); x++) {
-          for (int y = 1; y <= MapNumBadChanFull[test]->GetYaxis()->GetNbins(); y++) {
+        for (int x = 1; x <= MapNumBadChanFull[test]->GetXaxis()->GetNbins(); ++x) {
```
Comment (tjavaid) [inline DPGAnalysis/HcalTools/macros/cmt/RemoteMonitoringGLOBAL.cc:697]: why a pre-increment is now essential here?

**Comment by zhokin2** [comment]
I remember a few years ago I saw a little comment from a C++ guy about the correct use of ++i in loops rather than i++, so seeing it now I decided to fix it, although in principle, to be honest, I think that it doesn't really matter

________________________________________
От: Tahir Javaid ***@***.***>
Отправлено: 3 апреля 2024 г. 21:55
Кому: cms-sw/cmssw
Копия: Alexandr Zhokin; Mention
Тема: Re: [cms-sw/cmssw] Monitoring of Hcal adc sticking in neighbouring TS (PR #44600)

@tjavaid commented on this pull request.

________________________________

In DPGAnalysis/HcalTools/macros/cmt/RemoteMonitoringGLOBAL.cc<https://github.com/cms-sw/cmssw/pull/44600#discussion_r1550293215>:

> @@ -692,8 +694,8 @@ int main(int argc, char *argv[]) {
         MapNumBadChanDepth[test][sub][k]->Divide(
             MapNumBadChanDepth[test][sub][k], MapNumChanDepth[test][sub][k], 1, 1, "B");

-        for (int x = 1; x <= MapNumBadChanFull[test]->GetXaxis()->GetNbins(); x++) {
-          for (int y = 1; y <= MapNumBadChanFull[test]->GetYaxis()->GetNbins(); y++) {
+        for (int x = 1; x <= MapNumBadChanFull[test]->GetXaxis()->GetNbins(); ++x) {


why a pre-increment is now essential here?

**Comment by tjavaid** [comment]
please test

**Comment by zhokin2** [comment]
what you are waiting for

**Comment by tjavaid** [comment]
+1

**Comment by rappoccio** [comment]
+1

**Pull request merged**